### DISCUSSION TITLE
Bn Learner?

### DISCUSSION THREAD

**Discussion by kenneth-lee-ch** [agrumery/aGrUM#28] [Q&A] [answered]
Hi, if I have some discrete dataset, I want to impose a DAG structure on this dataset and and refill the potential based on the observed data and be able to derive the joint distribution according to factorization so that I can do a posterior update with some prior...I remember bnlearner is something relevant, but I am not sure. Would you please provide some pointers?

**Comment by phwuil** [comment, marked as answer]
Hi @kenneth-lee-ch , sorry or the delay :-)

When you know the structure and just want to "learn" the parameters from a database, the functions is `learnParameters`, you can choose a priori (for instance, a smoothing).

https://pyagrum.readthedocs.io/en/1.17.2/BNLearning.html#pyAgrum.BNLearner.learnParameters

https://pyagrum.readthedocs.io/en/1.17.2/notebooks/12-Examples_KaggleTitanic.html#Making-a-BN-without-learning-data-(structureFromExpert4)

**Comment by kenneth-lee-ch** [comment]
If I just want to count the frequency to fill the CPT, is learnParameters() the right option?

**Reply by phwuil** [reply]
Yes, with no prior (`learner.useNoPrior()`)... But beware of 0s  :-)

**Reply by kenneth-lee-ch** [reply]
How can I use fitParameters() and get back a BayesNet as a returned object at the same time like `learnParameters()`?

**Reply by phwuil** [reply]
You're right, I'm sorry for the confusion. It's better to use fitParameters than learnParameters (which should be deprecated...). learnParameters can be misleading, due to possible confusion between nodeids and column ids in the csv.

So the right way to adapt parameters of a BN using frequency of the databse :

```python
# bn is the BN to fit
learner=gum.BNLearner(csvfile,bn) # bn as the template in order to correctly parse the variable types.
learner.useNoPrior() # pure Max Likelihood => frequency
learner.fitParameters(bn)
```

**Reply by kenneth-lee-ch** [reply]
@phwuil  How can I get back the updated bn from `learner.fitParameters(bn)`?

**Reply by phwuil** [reply]
This is in place : `fitParameters(bn)` change the CPTs inside `bn`

**Reply by kenneth-lee-ch** [reply]
@phwuil I have a dataframe filled with discrete data and I have a bn structure that I created. When I ran this code `gum.BNLearner(df, bn)`, it gives me this error. I suspect the error means there is no such state in the current bn object, but I thought the function only takes the bn structure instead of storing the states? How can I resolve this issue?

```
learner = gum.BNLearner(df, bn)
pyAgrum.pyAgrum.UnknownLabelInDatabase: [pyAgrum] Unknown label found in database: The translation of String "3" could not be found for variable '1:Range([0,1])'
```

**Reply by kenneth-lee-ch** [reply]
I did this instead and the error is gone:
```
learner = gum.BNLearner(df)
learner.useSmoothingPrior()
 learned_bn = learner.learnParameters(bn.dag())
```

**Reply by phwuil** [reply]
Hi @kenneth-lee-ch ,
The problem seems to be that the BN you use as a template and container for the frequencies you want to calculate from the database is not correcty defined:

```
[pyAgrum] Unknown label found in database: The translation of String "3" could not be found for variable '1:Range([0,1])'
```

When creating the variable '1', you specify it as binary (by default I guess), but this column can contain some '3's ...

In order to correctly create your template, you can use a `pyAgrum.Discretizer` which parses the database and infers the node types:

![image](https://github.com/user-attachments/assets/7076bad7-e235-4125-b835-4834144700b6)

**Reply by kenneth-lee-ch** [reply]
@phwuil I tried your code with a randomly generated data [test.csv](https://github.com/user-attachments/files/18482845/test.csv)
, but I still get the error.

```
import pyAgrum as gum
import pandas as pd
df = pd.read_csv('test.csv', index_col=0)
edges = [(0, 1), (2, 0), (2, 1), (2, 8), (3, 2), (3, 4), (3, 6), (3, 8), (6, 7), (6, 8), (5, 6), (9, 5), (10, 0)]
from pyAgrum.lib.discretizer import Discretizer
discretizer = Discretizer()
bn = discretizer.discretizedTemplate(df)
for edge in edges:
    u, v, = edge
    bn.addArc(str(u), str(v))

learner = gum.BNLearner(df, bn)
learner.useNoPrior()
learner.fitParameters(bn)
```

Error message:
```
DatabaseError: [pyAgrum] Database error: The conditioning set <10=1, 2=1> for target node 0 never appears in the database. Please consider using priors such as smoothing.

```

**Reply by kenneth-lee-ch** [reply]
The dataset only has a single row with the column '10' being 1, so I don't follow what the fitParaemter() is looking up for the combination <10=1, 2=1>.

 <img width="360" alt="Screenshot 2025-01-20 at 5 22 50 PM" src="https://github.com/user-attachments/assets/51d7817d-6dc0-4fe7-9328-017712dcb30b" />

**Reply by kenneth-lee-ch** [reply]
![Screenshot** 2025-01-21 at 12 47 44 AM](https://github.com/user-attachments/assets/2522505e-7568-4940-a72d-b2ca00bd0f0c)
Another issue I have is the memory usage...as I repeatedly call `BNLearner`, it seems to stack up a lot of hard disk space for some reason. Specifcally, I mean this line 
```
learner = gum.BNLearner(joint_df, bn)
```
Would you mind to explain why even if I constantly replace the BNLearner with a different BN?
